<!DOCTYPE html>
<html>
  <head>
    <title>PEARG Documentation</title>
    
      <meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1, user-scalable=no">
<meta name="revised" content="2022-01-25T16:19:04 AEDT">
<title>Directory Structure :: PEARG Documentation</title>
<link rel="shortcut icon" href="/pearg_documentation/images/favicon.png" type="image/x-icon" />
<link href="/pearg_documentation/css/nucleus.css" rel="stylesheet">
<link href="/pearg_documentation/css/font-awesome.min.css" rel="stylesheet">
<link href="/pearg_documentation/css/featherlight.min.css" rel="stylesheet">
<link href="/pearg_documentation/css/auto-complete.css" rel="stylesheet">
<link href="/pearg_documentation/theme-flex/style.css" rel="stylesheet">

	<link href="/pearg_documentation/theme-flex/variant-blue.css" rel="stylesheet">

<link rel="stylesheet" href="/pearg_documentation/css/bootstrap.min.css">
<script>
  (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
  (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
  m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
  })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');

  ga('create', 'UA-79101-13', 'auto');
  ga('send', 'pageview');

</script>

    
  </head>
  <body data-url="/pearg_documentation/best-practices/directory-structure/">
    
      <header>
  <div class="logo">
    
	
  
    <p>PEARG Documentation</p>

  


  </div>
  <div class="burger"><a href="javascript:void(0);" style="font-size:15px;">&#9776;</a></div>
    <nav class="shortcuts">
            <li class="" role="">
                <a href="https://github.com/pearg/pearg_documentation"  rel="noopener">
                  <i class='fa fa-github'></i> Github repo
                </a>
            </li>
            <li class="" role="">
                <a href="https://gohugo.io/"  rel="noopener">
                  <i class='fa fa-book'></i> Hugo Documentation
                </a>
            </li>
            <li class="" role="">
                <a href="https://github.com/vjeantet/hugo-theme-docdock"  rel="noopener">
                  <i class='fa fa-bookmark'></i> DocDock Theme
                </a>
            </li>
            <li class="" role="">
                <a href="https://blogs.unimelb.edu.au/pearg/"  rel="noopener">
                  <i class='fa fa-globe'></i> PEARG homepage
                </a>
            </li>
    </nav>
</header>
<article>
  <aside>
    <ul class="menu">
          <li data-nav-id="/pearg_documentation/" class="dd-item">
          <a href="/pearg_documentation/">
            <i class="fa fa-fw fa-home"></i>
          </a>
          </li>
    <li data-nav-id="https://pearg.github.io/pearg_documentation/getting-started/" class="dd-item alwaysopen haschildren
        ">
      <div>
      <a href="/pearg_documentation/getting-started/">Getting Started</a>
            <i class="fa fa-angle-down fa-lg category-icon"></i>
      </div>
        <ul>
      <li data-nav-id="https://pearg.github.io/pearg_documentation/getting-started/hpc-resources/" class="dd-item">
        <div>
          <a href="/pearg_documentation/getting-started/hpc-resources/">
            Compute Resources
          </a>
        </div>
    </li>
      <li data-nav-id="https://pearg.github.io/pearg_documentation/getting-started/getting-an-account/" class="dd-item">
        <div>
          <a href="/pearg_documentation/getting-started/getting-an-account/">
            Getting an Account
          </a>
        </div>
    </li>
      <li data-nav-id="https://pearg.github.io/pearg_documentation/getting-started/getting-help/" class="dd-item">
        <div>
          <a href="/pearg_documentation/getting-started/getting-help/">
            Getting Help
          </a>
        </div>
    </li>
        </ul>
    </li>
    <li data-nav-id="https://pearg.github.io/pearg_documentation/best-practices/" class="dd-item alwaysopen haschildren
        ">
      <div>
      <a href="/pearg_documentation/best-practices/">Best Practices</a>
            <i class="fa fa-angle-down fa-lg category-icon"></i>
      </div>
        <ul>
      <li data-nav-id="https://pearg.github.io/pearg_documentation/best-practices/directory-structure/" class="dd-item">
        <div>
          <a href="/pearg_documentation/best-practices/directory-structure/">
            Directory Structure
          </a>
        </div>
    </li>
      <li data-nav-id="https://pearg.github.io/pearg_documentation/best-practices/naming-files/" class="dd-item">
        <div>
          <a href="/pearg_documentation/best-practices/naming-files/">
            Naming Files
          </a>
        </div>
    </li>
      <li data-nav-id="https://pearg.github.io/pearg_documentation/best-practices/backing-up-your-data/" class="dd-item">
        <div>
          <a href="/pearg_documentation/best-practices/backing-up-your-data/">
            Backing Up Your Data
          </a>
        </div>
    </li>
        </ul>
    </li>
    <li data-nav-id="https://pearg.github.io/pearg_documentation/metadata-documentation/" class="dd-item alwaysopen haschildren
        ">
      <div>
      <a href="/pearg_documentation/metadata-documentation/">Metadata Documentation</a>
            <i class="fa fa-angle-down fa-lg category-icon"></i>
      </div>
        <ul>
      <li data-nav-id="https://pearg.github.io/pearg_documentation/metadata-documentation/metadata-overview/" class="dd-item">
        <div>
          <a href="/pearg_documentation/metadata-documentation/metadata-overview/">
            Metadata Overview
          </a>
        </div>
    </li>
      <li data-nav-id="https://pearg.github.io/pearg_documentation/metadata-documentation/using-mediaflux/" class="dd-item">
        <div>
          <a href="/pearg_documentation/metadata-documentation/using-mediaflux/">
            Using Mediaflux
          </a>
        </div>
    </li>
      <li data-nav-id="https://pearg.github.io/pearg_documentation/metadata-documentation/mediaflux-explorer/" class="dd-item">
        <div>
          <a href="/pearg_documentation/metadata-documentation/mediaflux-explorer/">
            Mediaflux Explorer
          </a>
        </div>
    </li>
      <li data-nav-id="https://pearg.github.io/pearg_documentation/metadata-documentation/sample-map/" class="dd-item">
        <div>
          <a href="/pearg_documentation/metadata-documentation/sample-map/">
            Sample Map
          </a>
        </div>
    </li>
        </ul>
    </li>
    <li data-nav-id="https://pearg.github.io/pearg_documentation/usage-documentation/" class="dd-item alwaysopen haschildren
        ">
      <div>
      <a href="/pearg_documentation/usage-documentation/">Usage Documentation</a>
            <i class="fa fa-angle-down fa-lg category-icon"></i>
      </div>
        <ul>
      <li data-nav-id="https://pearg.github.io/pearg_documentation/usage-documentation/running-jobs-with-slurm/" class="dd-item">
        <div>
          <a href="/pearg_documentation/usage-documentation/running-jobs-with-slurm/">
            Running Jobs with SLURM
          </a>
        </div>
    </li>
      <li data-nav-id="https://pearg.github.io/pearg_documentation/usage-documentation/using-vnc/" class="dd-item">
        <div>
          <a href="/pearg_documentation/usage-documentation/using-vnc/">
            Using VNC
          </a>
        </div>
    </li>
      <li data-nav-id="https://pearg.github.io/pearg_documentation/usage-documentation/installing-software/" class="dd-item">
        <div>
          <a href="/pearg_documentation/usage-documentation/installing-software/">
            Installing Software
          </a>
        </div>
    </li>
      <li data-nav-id="https://pearg.github.io/pearg_documentation/usage-documentation/stacks-databases/" class="dd-item">
        <div>
          <a href="/pearg_documentation/usage-documentation/stacks-databases/">
            Stacks Databases
          </a>
        </div>
    </li>
      <li data-nav-id="https://pearg.github.io/pearg_documentation/usage-documentation/rad-seq-pipeline/" class="dd-item">
        <div>
          <a href="/pearg_documentation/usage-documentation/rad-seq-pipeline/">
            Using the RAD-seq Pipeline
          </a>
        </div>
    </li>
        </ul>
    </li>
    <li data-nav-id="https://pearg.github.io/pearg_documentation/tutorials/" class="dd-item alwaysopen haschildren
        ">
      <div>
      <a href="/pearg_documentation/tutorials/">Tutorials</a>
            <i class="fa fa-angle-down fa-lg category-icon"></i>
      </div>
        <ul>
      <li data-nav-id="https://pearg.github.io/pearg_documentation/tutorials/intro-rad-seq/" class="dd-item">
        <div>
          <a href="/pearg_documentation/tutorials/intro-rad-seq/">
            Introduction to RAD-Seq
          </a>
        </div>
    </li>
        </ul>
    </li>




    </ul>
    <section>
    </section>
  </aside>
  <section class="page">
    
    <div class="nav-select">
      <center>Navigation : 
        <select onchange="javascript:location.href = this.value;">
          
    <option value="/pearg_documentation/getting-started/" >
   Getting Started</option> 
      <option value="/pearg_documentation/getting-started/hpc-resources/" >- Compute Resources</option>
      <option value="/pearg_documentation/getting-started/getting-an-account/" >- Getting an Account</option>
      <option value="/pearg_documentation/getting-started/getting-help/" >- Getting Help</option>
  
    <option value="/pearg_documentation/best-practices/" >
   Best Practices</option> 
      <option value="/pearg_documentation/best-practices/directory-structure/"  selected>- Directory Structure</option>
      <option value="/pearg_documentation/best-practices/naming-files/" >- Naming Files</option>
      <option value="/pearg_documentation/best-practices/backing-up-your-data/" >- Backing Up Your Data</option>
  
    <option value="/pearg_documentation/metadata-documentation/" >
   Metadata Documentation</option> 
      <option value="/pearg_documentation/metadata-documentation/metadata-overview/" >- Metadata Overview</option>
      <option value="/pearg_documentation/metadata-documentation/using-mediaflux/" >- Using Mediaflux</option>
      <option value="/pearg_documentation/metadata-documentation/mediaflux-explorer/" >- Mediaflux Explorer</option>
      <option value="/pearg_documentation/metadata-documentation/sample-map/" >- Sample Map</option>
  
    <option value="/pearg_documentation/usage-documentation/" >
   Usage Documentation</option> 
      <option value="/pearg_documentation/usage-documentation/running-jobs-with-slurm/" >- Running Jobs with SLURM</option>
      <option value="/pearg_documentation/usage-documentation/using-vnc/" >- Using VNC</option>
      <option value="/pearg_documentation/usage-documentation/installing-software/" >- Installing Software</option>
      <option value="/pearg_documentation/usage-documentation/stacks-databases/" >- Stacks Databases</option>
      <option value="/pearg_documentation/usage-documentation/rad-seq-pipeline/" >- Using the RAD-seq Pipeline</option>
  
    <option value="/pearg_documentation/tutorials/" >
   Tutorials</option> 
      <option value="/pearg_documentation/tutorials/intro-rad-seq/" >- Introduction to RAD-Seq</option>
  



        </select>
      </center>
    </div>
      <div>
        <div class="searchbox">
          <input data-search-input id="search-by" type="text" placeholder="Search...">
        </div>
      </div>
    

    <h1>Directory Structure</h1>
    
    
    
    <p>When using the PEARG clusters or the Melbourne Bioinformatics clusters, your
files and data should be organised according to the convention described
here.</p>
<p>Each project should have its own directory located inside your home directory
or a shared location. Your project directory name should be meaningful.</p>
<p>Any shared data such as reference genomes and indices should be stored outside
your project directory to avoid unnecessary duplication of data.</p>

<div class="panel panel-warning"><div class="panel-heading">Check your project directory</div><div class="panel-body"><p>At absolute minimum, please check:</p>
<ul>
<li>is your project name sensible?</li>
<li>do you have a README file (or equivalent) in the parent directory of your
project directory that contains meaningful information?</li>
<li>do you have a data directory (containing your raw data) and a results
directory (containing processed data) with files organised in a logical
manner?</li>
</ul>
<p>If you&rsquo;ve answered no to any of these questions, you may get a stern email in
the future.</p>
</div></div>

<p>The motivation behind having a strict structure that all members adhere to
for project organisation is transparency. Anyone from the lab should be able
to look in your project directory and clearly understand what was run and
reproduce your analysis. Most of the time, the person trying to decode what
analyses were performed and why will be future you. Cooperate with your
future self by leaving verbose notes in README files!</p>
<p>While the sysadmin won&rsquo;t be authoritarian about the precise directory structure,
any flagrant disregard for the guidelines (such as dumping all your processing
files in the top-level results directory) will be met with consequences.</p>
<hr>
<h2 id="filesystem-overview">Filesystem overview</h2>
<p>Click on the image to view a larger version of the recommended directory
structure or scroll down for another example in text.</p>
<p><a href="https://swift.rc.nectar.org.au:8888/v1/AUTH_0edd079661d54bc0a0e4fde3b4892362/images/NGS_filesystem.png"><img src="https://swift.rc.nectar.org.au:8888/v1/AUTH_0edd079661d54bc0a0e4fde3b4892362/images/NGS_filesystem_thumbnail.png" alt=""></a></p>
<hr>
<h2 id="creating-a-project-directory-using-the-template">Creating a project directory using the template</h2>
<div class="highlight"><pre tabindex="0" style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4"><code class="language-bash" data-lang="bash"><span style="color:#75715e"># TODO: Jess will create a project skeleton in the future</span>
</code></pre></div><p>You should rename the directories starting with <code>rename_*</code> into something
sensible.</p>
<hr>
<h2 id="example-directory-structure">Example directory structure</h2>
<p>In this example, our project name is called <code>project_name</code> and is a RAD-seq
experiment that was processed using Stacks.</p>
<div class="highlight"><pre tabindex="0" style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4"><code class="language-text" data-lang="text">project_name/
├── data/
├── results/
├── scripts/
├── software/
└── README
</code></pre></div><p>The directory has four subdirectories: <code>data</code>, <code>results</code>, <code>scripts</code>, and
<code>software</code>, and one <code>README</code> file. The <code>README</code> file should be a plain-text
file containing basic project information (e.g. what the project is about,
what type of data was sequenced).</p>
<h3 id="data-directory">Data directory</h3>
<p>The <code>data</code> directory should contain the raw data received from sequencing.
Each library should have it&rsquo;s own directory containing sequencing files and
a text file containing barcodes corresponding to samples. This file is needed
for Stacks <code>process_radtags</code>.</p>
<div class="highlight"><pre tabindex="0" style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4"><code class="language-text" data-lang="text">project_name/
├── data/
│   ├── library_1_raw_data/
│   │   ├── seqA_R1_001.fastq.gz
│   │   ├── seqA_R2_001.fastq.gz
│   │   └──  library_1_barcodes.txt
│   ├── library_2_raw_data/
│   │   ├── seqB_R1_001.fastq.gz
│   │   ├── seqB_R2_001.fastq.gz
│   │   └── library_2_barcodes.txt
├── results/
├── scripts/
├── software/
└── README
</code></pre></div><h4 id="make-files-read-only-optional">Make files read-only (optional)</h4>
<p>The files in your data directory should never be edited.</p>
<p>If you are familiar with UNIX file permissions, you can remove write
permissions with the <code>chmod</code> command. For example, the
following command removes write permission for all users:</p>
<div class="highlight"><pre tabindex="0" style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4"><code class="language-bash" data-lang="bash">chmod a-w seqA_R1_001.fastq.gz
</code></pre></div><p>You can check file permissions with <code>ls -l</code> where the first column represents
whether read/write/execute access is avaiable.</p>
<div class="highlight"><pre tabindex="0" style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4"><code class="language-text" data-lang="text">$ ls -l
-r--r--r-- 1 jess jess 142870 Aug  8 14:30 seqA_R1_001.fastq.gz
-r--r--r-- 1 jess jess 177552 Aug  8 14:30 seqA_R2_001.fastq.gz
</code></pre></div><h3 id="results-directory">Results directory</h3>
<p>The <code>results</code> directory should have one directory for each time you generate
a set of results. Using subdirectories inside the main <code>results</code> directory is
recommended because often experiments are re-run in the future (e.g. updated
software versions, more sequencing data, reanalysis before publication).
I recommend naming the directory with a date in <code>YYYY-MM-DD</code> format at the
beginning of the name so the directories are sorted chronologically.</p>
<div class="highlight"><pre tabindex="0" style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4"><code class="language-text" data-lang="text">project_name/
├── data/
├── results/
│   ├── 2017-08-01_results/
│   │   ├── ...
│   │   ├── ...
│   │   └── ...
│   ├── 2017-11-01_extra_samples/
│   │   ├── ...
│   │   ├── ...
│   │   └── ...
│   ├── 2018-05-01_reanalysis/
│   │   ├── ...
│   │   ├── ...
│   │   └── ...
├── scripts/
├── software/
└── README
</code></pre></div><p>Inside each result subdirectory, there should be multiple directories
containing output from steps in your workflow. In this example, the directories
inside <code>2017-08-01_results</code> are: <code>demuxed_seq</code>, <code>demuxed_cat</code>, <code>alignments</code>,
<code>stacks</code>, and <code>qc</code>. Your directory names may look different depending on what
type of analysis you&rsquo;re performing. The contents of each directory is described
below.</p>
<h4 id="sequencing-data">Sequencing data</h4>
<div class="highlight"><pre tabindex="0" style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4"><code class="language-text" data-lang="text">results/
├── 2017-08-01_results/
│   ├── demuxed_seq/
│   │   ├── mozzie-1.1.fq
│   │   ├── mozzie-1.2.fq
│   │   ├── mozzie-1.rem.1.fq
│   │   ├── mozzie-1.rem.2.fq
│   │   ├── mozzie-2.1.fq
│   │   ├── mozzie-2.2.fq
│   │   └── ...
│   ├── demuxed_cat/
│   │   ├── mozzie-1.fq
│   │   ├── mozzie-2.fq
│   │   ├── mozzie-3.fq
│   │   └── ...
│   └── ...
</code></pre></div><p>The <code>demuxed_seq</code> directory contains demuxed sequencing data processed by
<code>process_ragtags</code>. Stacks should output four files for each sample listed in
the barcode file. In this example, <code>mozzie-1.1.fq</code> and <code>mozzie-1.2.fq</code> contain
the set forward and reverse reads for the <code>mozzie-1</code> sample. The
<code>mozzie-1.rem.1.fq</code> and <code>mozzie-1.rem.2.fq</code> files contain the remaining reads
that are unpaired due to their mate being discarded.</p>
<p>If you&rsquo;re working with ddRADseq data, Stacks recommends concatenating the four
files together. Here, <code>demuxed_cat</code> contains the concatenated files.</p>
<h4 id="alignment-data">Alignment data</h4>
<div class="highlight"><pre tabindex="0" style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4"><code class="language-text" data-lang="text">results/
├── 2017-08-01_results/
│   ├── demuxed_seq/
│   ├── demuxed_cat/
│   ├── alignments/
│   │   ├── AaegL2/
│   │   │   ├── mozzie-1.AaegL2.sorted.bam
│   │   │   ├── mozzie-1.AaegL2.sorted.bam.bai
│   │   │   ├── mozzie-2.AaegL2.sorted.bam
│   │   │   ├── mozzie-2.AaegL2.sorted.bam.bai
│   │   │   ├── ...
│   │   │   └── AagL2_alignment_code.txt
│   │   ├── AaegL3/
│   │   │   ├── mozzie-1.AaegL3.sorted.bam
│   │   │   ├── mozzie-1.AaegL3.sorted.bam.bai
│   │   │   ├── mozzie-2.AaegL3.sorted.bam
│   │   │   ├── mozzie-2.AaegL3.sorted.bam.bai
│   │   │   ├── ...
│   │   │   └── AagL3_alignment_code.txt
│   └── ...
</code></pre></div><p>Alignments should be stored in the <code>alignments</code> directory with a separate
directory for each reference genome aligned against. Alignments should be stored
as bam files with the <code>.bam</code> suffix and bam index files, if provided, should end
with <code>.bai</code>. If alignements are sorted, it&rsquo;s recommended to include <code>sorted</code> in
the filename. Including the reference genome name in the filename is also
helpful.</p>
<p>A plain-text file with what commands were run should also be included in the
directory (e.g. <code>AagL2_alignment_code.txt</code>) or in the <code>scripts</code> directory.</p>
<h4 id="stacks-data">Stacks data</h4>
<div class="highlight"><pre tabindex="0" style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4"><code class="language-text" data-lang="text">results/
├── 2017-08-01_results/
│   ├── demuxed_seq/
│   ├── demuxed_cat/
│   ├── alignments/
│   ├── stacks/
│   │   ├── stacks_AaegL2_females/
│   │   │   ├── catalog/
│   │   │   │   ├── mozzie-1.AaegL2.alleles.tsv
│   │   │   │   ├── mozzie-1.AaegL2.matches.tsv
│   │   │   │   ├── mozzie-1.AaegL2.models.tsv
│   │   │   │   ├── mozzie-1.AaegL2.snps.tsv
│   │   │   │   ├── mozzie-2.AaegL2.alleles.tsv
│   │   │   │   ├── ...
│   │   │   │   ├── batch_1.catalog.alleles.tsv
│   │   │   │   ├── batch_1.catalog.snps.tsv
│   │   │   │   ├── batch_1.catalog.tags.tsv
│   │   │   │   └── batch_1.markers.tsv
│   │   │   ├── population_females_filtered/
│   │   │   │   ├── batch_1.vcf
│   │   │   │   ├── batch_1.haplotypes.tsv
│   │   │   │   ├── batch_1.sumstats_summary.tsv
│   │   │   │   ├── batch_1.sumstats.tsv
│   │   │   │   ├── batch_1.hapstats.tsv
│   │   │   │   └── code_females_filtered.txt
│   │   │   ├── population_females_singlesnp/
│   │   │   │   ├── batch_1.vcf
│   │   │   │   ├── batch_1.haplotypes.tsv
│   │   │   │   ├── batch_1.sumstats_summary.tsv
│   │   │   │   ├── batch_1.sumstats.tsv
│   │   │   │   ├── batch_1.hapstats.tsv
│   │   │   │   └── code_females_singlesnp.txt
│   │   ├── stacks_AaegL2_males/
│   │   │   └── ...
│   │   └── ...
│   └── ...
</code></pre></div><p>Each run of Stacks to get a catalogue should have its own separate directory
in the <code>stacks</code> directory. The output files from <code>ref_map</code> or <code>denovo_map</code>
stored in its own directory.</p>
<p>Each time you run Stacks <code>populations</code> with designated filters, you should
store the files in a separate directory. You should also include a file
containing the code that was used to produce the output.</p>
<h4 id="qc-data">QC data</h4>
<div class="highlight"><pre tabindex="0" style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4"><code class="language-text" data-lang="text">results/
├── 2017-08-01_results/
│   ├── demuxed_seq/
│   ├── demuxed_cat/
│   ├── alignments/
│   ├── stacks/
│   ├── qc/
│   │   ├── fastqc/
│   │   │   ├── ...
│   │   │   ├── ...
│   │   │   └── ...
│   │   ├── flagstat/
│   │   │   ├── ...
│   │   │   ├── ...
│   │   │   └── ...
│   │   └── ...
│   └── ...
</code></pre></div><p>The <code>qc</code> directory should contain the output of programs run for quality
control purposes (e.g. fastQC, samtools flagstat).</p>
<h3 id="scripts-directory">Scripts directory</h3>
<p>It&rsquo;s up to you if you want to store your scripts inside the <code>scripts</code> directory
or with the output files that were generated. Just make sure you document
all the code that was run somewhere sensible.</p>
<h3 id="software-directory">Software directory</h3>
<p>If you have any additional software you compiled specifically for your project,
you can store them here.</p>
<h3 id="readme-files">README files</h3>
<p>README files are plain-text files where you should write descriptions of what
the directory contains, what analysis was done, why certain parameters were
chosen, what results were found, etc. Place a README file in any directory
you feel could use one. Documenting your work clearly is good practice and
often pays dividends in the future.</p>


    
    
        <div class="chevrons">
    <div id="navigation">
<a class="nav nav-prev" href="/pearg_documentation/best-practices/" title="Best Practices"> <i class="fa fa-chevron-left"></i><label>Best Practices</label></a>
    <a class="nav nav-next" href="/pearg_documentation/best-practices/naming-files/" title="Naming Files" style="margin-right: 0px;"><label>Naming Files</label><i class="fa fa-chevron-right"></i></a></div>
  </div>

  </section>
</article>

<footer>

<div class="footline">
    

    

    
    <div class="date">
        <i class='fa fa-calendar'></i> Last update on 08/08/2017
    </div>
    

    
    <div class="github-link">
      <a href="https://github.com/pearg/pearg_documentation/edit/master/content/best-practices/directory-structure.md" target="blank"><i class="fa fa-code-fork"></i>
        Improve this page</a>
    </div>
    
  </div>


	<div>


  
    Create a content/_footer.md file to customize the footer content
  



	</div>
</footer>

<script src="/pearg_documentation/js/jquery-2.x.min.js"></script>
<script type="text/javascript">
    
        var baseurl = "https:\/\/pearg.github.io\/pearg_documentation\/";
    
</script>
<script src="/pearg_documentation/js/clipboard.min.js"></script>
<script src="/pearg_documentation/js/featherlight.min.js"></script>

<script type="text/javascript" src="/pearg_documentation/js/lunr.min.js"></script>
<script type="text/javascript" src="/pearg_documentation/js/auto-complete.js"></script>
<script type="text/javascript" src="/pearg_documentation/js/search.js"></script>
<script src="/pearg_documentation/theme-flex/script.js"></script>




<script src="https://pearg.github.io/pearg_documentation/js/scripts.min.js"></script>
    

    
    

    
  </body>
</html>